### PR TITLE
fixed some compiler errors under gcc and clang

### PR DESCRIPTION
**Pull request by georg-emg**
Fixed a number of compiler errors that occurred when com piling under Linux. Tested were g++-12 and clang++-14 under Debian Bookworm.

### CODE REVIEW

**Comment by georg-emg** [comment]
The changes in move_only_function.h are quite hacky. Maybe a better solution is possible.

File: CMakeLists.txt [hunk 1]
```diff
@@ -35,6 +35,7 @@ else()
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wsign-conversion>)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wsign-compare>)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-fno-rtti>)
+    add_compile_options($<$<COMPILE_LANG_AND_ID:CXX,Clang>:-fsized-deallocation>)
```
Comment (zhihaoy) [inline CMakeLists.txt:38]: What errors did you see without this option?

File: tests/include/boost/ut.hpp [hunk 2]
```diff
@@ -331,28 +331,28 @@ struct identity {
 };
 
 template <class T>
-struct function_traits : function_traits<decltype(&T::operator())> {};
+struct function_trait : function_trait<decltype(&T::operator())> {};
```
Comment (zhihaoy) [inline tests/include/boost/ut.hpp:334]: BoostExt.ut is tested under gcc 9 and clang 9. What's the problem with this name?

File: tests/move_only_function/test_noexcept.cpp [hunk 3]
```diff
@@ -7,6 +7,7 @@ constexpr auto nothrow_call = [](move_only_function<int() noexcept> f)
 
 template<class... T> struct overload : T...
 {
+    constexpr overload(const T &... t) : T(t)... {}
```
Comment (zhihaoy) [inline tests/move_only_function/test_noexcept.cpp:10]: How about adding a deduction guide
```cpp
template<class... T>
overload(T...) -> overload<T...>;
```
after the class body as a workaround?

File: tests/function/test_nontype.cpp [hunk 4]
```diff
@@ -8,6 +8,9 @@ namespace
 
 struct A
 {
+    A() = default;
+    A(int a) : val(a) {}
```
Comment (zhihaoy) [inline tests/function/test_nontype.cpp:12]: What errors did you see?

**Review by zhihaoy — COMMENTED** [review]
It's possible that gcc-12 introduced some regression. If you found any, please share so that I can ensure the regressions have been reported.

Nevertheless, these changes had better be maintained separately. How about this?
1. I create a new branch called `prod` or something you can recognize;
2. You can target the PR to that branch;
3. I setup Clang and maybe more compilers in GitHub Actions to ensure that future changes do not regress;
4. New changes will be ported to that branch once stable.

File: tests/include/boost/ut.hpp (hunk 2 above)
Comment (georg-emg) [reply tests/include/boost/ut.hpp]: Sorry, that was a global search/replace error. I reverted that change.

File: tests/move_only_function/test_noexcept.cpp (hunk 3 above)
Comment (georg-emg) [reply tests/move_only_function/test_noexcept.cpp:11]: This is probably a bug in clang/clang-cl (observed in versions 14 and 15). Clang gives the following error:

~~~
error: no viable constructor or deduction guide for deduction of template arguments of 'overload'
~~~

Apparrently, it does not consider the member-wise struct initializer when deducinf the template arguments. GCC handles the original code just fine.

I wrapped the workarounf in #ifdef __clang__.

File: tests/function/test_nontype.cpp (hunk 4 above)
Comment (georg-emg) [reply tests/function/test_nontype.cpp:13]: This is the same clang problem as above, only this time is gives the following error:

~~~
error: no matching function for call to 'construct_at'

tests/function/test_nontype.cpp:109:41: note: in instantiation of function template specialization 'std::make_shared<(anonymous namespace)::A, int>' requested here
                        auto ptr = std::make_shared<A>(6);
~~~

Again, it fails to find the memberwise constructor of the struct. I also wrapped this in `#ifdef __clang__`

File: CMakeLists.txt (hunk 1 above)
Comment (georg-emg) [reply CMakeLists.txt:38]: Clang does not support sized deallocation unless -fsized-deallocation is specified. This results in the following error:

~~~
tests/move_only_function/test_unique.cpp:48:9: error: no matching function for call to 'operator delete'
        ::operator delete(s, full_size);
        ^~~~~~~~~~~~~~~~~
~~~

Either the flag must be specified, or the relevant lines must be wrapped in `#ifdef __cpp_sized_deallocation`.

**Comment by georg-emg** [comment]
If you want me to target the PR to a different branch, just tell me whioch branch you would like. Or you can retarget it yourself, I think.

**Comment by georg-emg** [comment]
I converted the pull request to a draft, because I am not happy with what I did in std23::move_only_function. I'll clean up the code and re-publish it.

**Comment by georg-emg** [comment]
Ok, I refactored std23::move_only_function to avoid using static_cast. The PR is ready again now.

File: CMakeLists.txt (hunk 1 above)
Comment (zhihaoy) [reply CMakeLists.txt:38]: Because this is in tests, I'm going to enable the compiler option instead.

File: tests/function/test_nontype.cpp (hunk 4 above)
Comment (zhihaoy) [reply tests/function/test_nontype.cpp:13]: Ah, I forgot this one. It's a different problem, though. I'll explain later.

File: tests/move_only_function/test_noexcept.cpp (hunk 3 above)
Comment (zhihaoy) [reply tests/move_only_function/test_noexcept.cpp:11]: As the error message said, missing the (implicitly generated) deduction guide was the problem.

**Review by zhihaoy — COMMENTED** [review]
I think it doesn't make sense to maintain a separate copy of tests, and it doesn't worth the effort to add conditional compilation there. So, I committed the test-related changes 83ba1aeb to `main`, and branched off `compat`. I've retargeted your branch on top of `compat`. Please revert the changes I commented on, and try CI again.

File: include/std23/move_only_function.h [hunk 5]
```diff
@@ -301,24 +301,178 @@ template<class S, class = typename _full_fn_sig<S>::function>
...
+    {
+        return call(std::forward<Args>(args)...);
+    }
+};
+
+template<class S, class R, class... Args>
+class move_only_function<S, R(Args...)> : _move_only_function_base<S, R, Args...>
+{
+    using base = _move_only_function_base<S, R, Args...>;
+    friend base;
```
[the remark this answers is not in the document]
Comment (georg-emg) [reply include/std23/move_only_function.h]: Yes, this is a left-over, I removed it.

File: include/std23/move_only_function.h [hunk 6]
```diff
@@ -301,24 +301,178 @@ template<class S, class = typename _full_fn_sig<S>::function>
...
+    }
+};
+
+template<class S, class R, class... Args>
+    requires _move_only_function_base_base<S, R, Args...>::is_rvalue_only
+struct _move_only_function_base<S, R, Args...> : _move_only_function_base_base<S, R, Args...>
+{
+    using base = _move_only_function_base_base<S, R, Args...>;
+
+    using typename base::signature;
```
[the remark this answers is not in the document]
Comment (georg-emg) [reply include/std23/move_only_function.h]: I removed the using for the signature, as it is unused.

File: include/std23/move_only_function.h [hunk 7]
```diff
@@ -301,24 +301,178 @@ template<class S, class = typename _full_fn_sig<S>::function>
...
+        return vtbl_.get().call(obj_.val, std::forward<Args>(args)...);
+    }
+};
+
+template<class S, class R, class... Args>
+struct _move_only_function_base : _move_only_function_base_base<S, R, Args...>
+{
+    using base = _move_only_function_base_base<S, R, Args...>;
+
+    using typename base::signature;
```
[the remark this answers is not in the document]
Comment (georg-emg) [reply include/std23/move_only_function.h]: ditto.

**Comment by georg-emg** [comment]
I rebased my branch onto compat, to make merging easier. All changes to the tests were removed, the tests are no longer affected by the PR.

File: include/std23/move_only_function.h [hunk 8]
```diff
@@ -475,41 +623,9 @@ class move_only_function<S, R(Args...)>
...
-    {
-        return vtbl_.get().call(obj_.val, std::forward<Args>(args)...);
-    }
-
-    R operator()(Args... args) const &&noexcept(noex)
-        requires(is_const and !is_lvalue_only and is_rvalue_only)
-    {
-        return vtbl_.get().call(obj_.val, std::forward<Args>(args)...);
-    }
+  private:
```
[the remark this answers is not in the document]
Comment (georg-emg) [reply include/std23/move_only_function.h]: Removed the `private`

**Review by zhihaoy — APPROVED** [review]
Thank you for the work!

**Pull request merged**